mlir/include/mlir/Dialect/Ptr/IR/MemorySpaceInterfaces.td

```diff
@@ -0,0 +1,117 @@
+//===-- MemorySpaceInterfaces.td - Memory space interfaces ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines memory space attribute interfaces.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef PTR_MEMORYSPACEINTERFACES
+#define PTR_MEMORYSPACEINTERFACES
+
+include "mlir/IR/AttrTypeBase.td"
+include "mlir/IR/OpBase.td"
+
+//===----------------------------------------------------------------------===//
+// Memory space attribute interface.
+//===----------------------------------------------------------------------===//
+
+def MemorySpaceAttrInterface : AttrInterface<"MemorySpaceAttrInterface"> {
+  let description = [{
+    This interface defines a common API for interacting with the memory model of
+    a memory space and the operations in the pointer dialect.
+
+    Furthermore, this interface allows concepts such as read-only memory to be
+    adequately modeled and enforced.
+  }];
+  let cppNamespace = "::mlir::ptr";
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/        [{
+        This method checks if it's valid to load a value from the memory space
+        with a specific type, alignment, and atomic ordering.
+        If `emitError` is non-null then the method is allowed to emit errors.
+      }],
+      /*returnType=*/  "::mlir::LogicalResult",
```

Comment:
@fabianmcg : nitpicking here but methods that starts with `is` are predicates and should return `bool`: `LogicalResult` is a about method that can fail instead. Can you update?

Reply:
I can update. But these methods can emit an error, remember they get a `::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError` parameter.

Admittedly,  I would prefer these methods to return something like a `SuccesOrDiagnostic`, so users can have more control on how to handle the result. Or to decouple the diagnostic from the validity method.

Reply:
`FailureOr<bool>`?

Reply:
That's a good immediate option, I'll change to that. But I think we are missing something similar to `llvm::Expected`, like:

```c++
template <typename T>
class DiagnosticOr : public std::optional<T> {
   operator LogicalResult () const;
   Diagnostic takeDiagnostic();
   void emitDiagnostic();
   T& getValue();
private:
  Diagnostic message;
};
```
Which would allow users to decide how to handle the diagnostic.

Reply:
I think we're trying to be more efficient by avoiding formatting a diagnostic when the user does not want it.

Reply:
On success we shouldn't format a diagnostic, but I'm still not convinced about the approach.
I'll change to `FailureOr<bool>`, I'll revisit if I have a better idea.

Reply:
> On success we shouldn't format a diagnostic

I'm talking about the failing case where the user of the API would just discard the diagnostic.

Reply:
I just remembered  the `emitError` is always optional and the user of the API can opt to pass a default constructed object, in which case the method cannot produce errors. Therefore, it already handles the case of discarding the diagnostic in a efficient manner.

I'll switch to `FailureOr<bool>`.